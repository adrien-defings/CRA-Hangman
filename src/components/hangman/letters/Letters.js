import React from 'react'
import PropTypes from 'prop-types'
import {ALPHABET} from '../../utils.js'

const Letters = ({ usableLetters, color, onClick }) => {
    return (
        <div id="letters">
          <h5>
            Use
            <span className={"text-" + color}> keyboards </span>
            or
            <span className={"text-" + color}> letter buttons : </span>
          </h5>
          {
            usableLetters.map((letter, index) => (
              <span
              key={index}
              className={"letter btn btn-" + color}
              onClick={() => onClick(letter)}>
                {letter}
              </span>
            ))
          }
        </div>
    )
}

Letters.defaultProps = {
    usableLetters: ALPHABET,
    color: "primary",
    onClick: () => {},
}

Letters.propTypes = {
    usableLetters: PropTypes.arrayOf(PropTypes.string).isRequired,
    color: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
}

export default Letters