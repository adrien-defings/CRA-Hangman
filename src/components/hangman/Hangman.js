
import React, { Component } from 'react'
import PropTypes from 'prop-types'

import randomWords from 'random-words'
import clonedeep from 'lodash.clonedeep'

// eslint-disable-next-line
import { ALPHABET, WORDS } from '../utils.js'
import './hangman.css'
import InfosJoueur from './infosJoueur/InfosJoueur.js'
import HighScoreInput from './hallOfFame/HighScoreInput'
import HallOfFame from './hallOfFame/HallOfFame'
import Letters from './letters/Letters.js'

export default class Hangman extends Component {

  constructor(props) {
    super(props)
    this.state = {
      hiddenWord: this.getHiddenWord(),
      usableLetters: clonedeep(ALPHABET),
      currentPlayer: 0,
      movesCount: [0, 0],
      hallOfFame: null,
    }
    this.INITIAL_STATE = clonedeep(this.state)
  }

  getHiddenWord() {
    // return ["T","E","S","T"]

    // return english word
    return randomWords().toUpperCase().split('')

    // return french word
    // return WORDS[Math.floor(Math.random() * WORDS.length)].split('')
  }

  getHangmanStep(usableLetters, usedLetters) {
    let { hiddenWord } = this.state
    // Getting found hidden word letters
    let foundLetters = this.getWord(usedLetters, hiddenWord)
    foundLetters = foundLetters.filter((letter, index) => {
      return foundLetters.indexOf(letter) === index && letter !== "_"
    })

    // ALPHABET letters number - USABLE letters number ( = USED letters ) - FOUND letters = FAILED moves
    // = canvas/hangman displaying step
    return (ALPHABET.length - usableLetters.length) - foundLetters.length
  }

  isGameEnded(usableLetters) {
    let { hiddenWord } = this.state
    let usedLetters = this.getUsedLetters(usableLetters)
    return !this.getWord(usedLetters, hiddenWord).includes("_") || this.getHangmanStep(usableLetters, usedLetters) > 9
  }

  getWinner(hiddenWord, currentPlayer, usedLetters) {
    return !this.getWord(usedLetters, hiddenWord).includes("_") ? currentPlayer : 3
  }

  // Give a textual representaion of game state
  // Each hidden letters are represented by _underscore_
  getWord(usedLetters, hiddenWord) {
    return hiddenWord.map((letter) => (usedLetters.includes(letter) ? letter : '_'))
  }

  getUsedLetters(usableLetters) {
    return ALPHABET.filter((letter) => { return usableLetters.indexOf(letter) === -1 })
  }

  displayHallOfFame = (hallOfFame) => {
    this.setState({ hallOfFame })
  }

  drawCanvas(usableLetters, givenStep = null, canvasColor = null) {
    let usedLetters = this.getUsedLetters(usableLetters)
    let step = givenStep != null ? givenStep : this.getHangmanStep(usableLetters, usedLetters) - 1
    const canvas = document.getElementById('canvas')

    let ctx = canvas.getContext('2d')
    ctx.strokeStyle = canvasColor != null ? canvasColor : "#333"
    ctx.lineWidth = 2
    ctx.lineCap = "round"
    ctx.lineJoin = "round"

    const steps = [
      function step1() {
        ctx.strokeRect(30, 80, 40, 10)
      },

      function step2() {
        ctx.beginPath()
        ctx.moveTo(25, 90)
        ctx.lineTo(0, 90)
        ctx.stroke()
      },

      function step3() {
        ctx.beginPath()
        ctx.moveTo(10, 89)
        ctx.lineTo(10, 0)
        ctx.stroke()
      },

      function step4() {
        ctx.beginPath()
        ctx.moveTo(10, 1)
        ctx.lineTo(50, 1)
        ctx.stroke()
      },

      function step5() {
        ctx.beginPath()
        ctx.moveTo(50, 1)
        ctx.lineTo(50, 20)
        ctx.stroke()
      },

      function step6() {
        ctx.beginPath()
        ctx.arc(50, 30, 10, 0, Math.PI * 2, true)  // Cercle ext√©rieur
        ctx.stroke()
      },

      function step7() {
        ctx.beginPath()
        ctx.moveTo(50, 40)
        ctx.lineTo(50, 60)
        ctx.stroke()
      },

      function step8() {
        ctx.beginPath()
        ctx.moveTo(40, 50)
        ctx.lineTo(60, 50)
        ctx.stroke()
      },

      function step9() {
        ctx.beginPath()
        ctx.moveTo(50, 60)
        ctx.lineTo(40, 79)
        ctx.stroke()
        ctx.beginPath()
        ctx.moveTo(50, 60)
        ctx.lineTo(60, 79)
        ctx.stroke()
      },

      function step10() {
        ctx.clearRect(29, 79, 42, 12)
      },

    ]

    if (0 <= step && step < 10) steps[step]()

  }

  drawOpaqueHangman(usableLetters) {
    for (let i = 0; i < 9; i++) {
      this.drawCanvas(usableLetters, i, "#CCC")
    }
  }

  // fn arrow for binding
  reboot = () => {
    let state = clonedeep(this.INITIAL_STATE)
    state.hiddenWord = this.getHiddenWord()
    this.setState(state)

    // Canvas ( = hangman display ) cleaning
    const canvas = document.getElementById('canvas')
    const context = canvas.getContext('2d')
    context.clearRect(0, 0, 95, 95)

    this.drawOpaqueHangman(ALPHABET)
  }

  // fn arrow for binding
  onLetterClick = letter => {
    let { hiddenWord, usableLetters, movesCount, currentPlayer } = this.state
    let navigationMode = this.props.navigationMode

    // usable letters and current player moves count update
    usableLetters.splice(usableLetters.indexOf(letter), 1)
    movesCount[currentPlayer] += 1

    // if failed move
    if (!hiddenWord.includes(letter)) {
      // Canvas ( = hangman display ) update
      this.drawCanvas(usableLetters)

      if (navigationMode === "multiplayer" && !this.isGameEnded(usableLetters)) {
        currentPlayer = currentPlayer === 0 ? 1 : 0
      }
    }

    this.setState({
      usableLetters,
      movesCount,
      currentPlayer,
    })
  }

  componentDidMount() {
    let { usableLetters } = this.state
    this.drawOpaqueHangman(usableLetters)

    document.addEventListener('keydown', (event) => {
      let { usableLetters } = this.state
      let key = event.key.toUpperCase()
      if (usableLetters.includes(key) && !this.isGameEnded(usableLetters)) this.onLetterClick(key)
    })

  }

  componentWillUnmount(){
    document.removeEventListener("keydown", {});
  }

  render() {
    let { hiddenWord, usableLetters, currentPlayer, movesCount, hallOfFame } = this.state
    let isGameEnded = this.isGameEnded(usableLetters)
    let usedLetters = this.getUsedLetters(usableLetters)
    let winner = this.getWinner(hiddenWord, currentPlayer, usedLetters)
    const colors = ["primary", "danger"]
    let navigationMode = this.props.navigationMode
    let word = isGameEnded ? hiddenWord : this.getWord(usedLetters, hiddenWord)

    return (
      <div id="gameContainer">

        <canvas height={95} width={95} id="canvas">
          Sorry, this browser does not support &ltcanvas&gt.
        </canvas>

        <InfosJoueur
          isGameEnded={isGameEnded}
          winner={winner}
          color={colors[currentPlayer]}
          navigationMode={navigationMode}
          currentPlayer={currentPlayer}
        />
        <span>{movesCount[currentPlayer]} moves</span>

        <div id="word">
          {
            word.map((letter, index) => {
              return (
                <span key={index} letter={letter} className="letter badge badge-light">
                  {letter}
                </span>
              )
            })
          }
        </div>

        {!isGameEnded ?
          <Letters
            usableLetters={usableLetters}
            color={colors[currentPlayer]}
            onClick={this.onLetterClick}
          />
          :
          <div id="successActions">
            <button onClick={this.reboot} className="btn btn-success">Play Again !</button>
            {
              winner !== 3 &&
              <div id="hallOfFame">
                <hr />
                {hallOfFame ?
                  <HallOfFame entries={hallOfFame} />
                  :
                  <HighScoreInput guesses={movesCount[currentPlayer]} word={word.toString().replace(/,/g, "")} onStored={this.displayHallOfFame} />
                }
              </div>
            }
          </div>
        }
      </div>
    )

  }
}

Hangman.defaultProps = {
  navigationMode: 'singlePlayer'
}

Hangman.propTypes = {
  navigationMode: PropTypes.string.isRequired,
}