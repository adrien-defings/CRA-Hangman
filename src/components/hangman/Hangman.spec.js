import React from 'react'

import { expect } from 'chai'
import { shallow } from 'enzyme'
import sinon from 'sinon'

import Hangman from './Hangman'
import Letters from './letters/Letters'
import { ALPHABET } from '../utils.js'

describe('<Hangman />', () => {

  let wrapper
  const navigationModes = ["singlePlayer", "multiplayer"]
  let navigationMode = navigationModes[0]

  // Render the component in the beforeEach
  // => Each test can get a new shallow rendered wrapper
  beforeEach(() => {

    // setup a DOM element as a render target
    const container = document.createElement("div")
    container.setAttribute("id", "hangmanTestContainer")
    document.body.appendChild(container)

    // create the canvas dom element
    document.querySelector("#hangmanTestContainer").innerHTML = `
      <canvas id="canvas">
        Sorry, this browser does not support &ltcanvas&gt.
        {console.log("Hm, something wents wrong on &ltcanvas&gt build in hangman.spec.js.")}
      </canvas>
    `

    navigationMode = navigationModes[Math.floor(Math.random() * navigationModes.length)]
    wrapper = shallow( <Hangman navigationMode={navigationMode} />)
  })

  it('should match with prev snapshot', () => {
    let wrapperSnapshot = shallow( <Hangman navigationMode="multiplayer" />)
    wrapperSnapshot.setState({ hiddenWord: ["T","E","S","T"] })
    expect(wrapperSnapshot).to.matchSnapshot()
  })

  // Checking the initial state values
  it('should initialize the hiddenWord state to a string', () => {
    expect(typeof expect(wrapper).to.have.state('hiddenWord')).to.be.a('string')
  })

  it('should initialize the usableLetters state to an array of Alphabet chars', () => {
    expect(wrapper).to.have.state('usableLetters').to.deep.equal(ALPHABET)
  })

  it('should initialize the currentPlayer state to 0', () => {
    expect(wrapper).to.have.state('currentPlayer').to.equal(0)
  })

  it('should initialize the movesCount state to [0, 0]', () => {
    expect(wrapper).to.have.state('movesCount').to.deep.equal([0, 0])
  })

  it('should initialize the hallOfFame state to null', () => {
    expect(wrapper).to.have.state('hallOfFame').to.equal(null)
  })

  // on letter click
  it('should trigger its `onClick` on letter Button', () => {
    const onClick = sinon.spy()
    const wrapper = shallow(
      <Letters usableLetters={ALPHABET} color="primary" onClick={onClick} />
    )
    const letterBtn = wrapper.find('.letter.btn').at(0)

    // expect(letterBtn).to.be.truthy
    letterBtn.simulate("click")
    expect(onClick).to.have.been.calledWith("A")
  })

  // onLetterClick state updating tests
  it('should update the state usableLetters to the correct value', () => {
    const char = ALPHABET[Math.floor(Math.random() * ALPHABET.length)]
    ALPHABET.splice(ALPHABET.indexOf(char), 1)
    wrapper.instance().onLetterClick(char)
    expect(wrapper).to.have.state('usableLetters').to.deep.equal(ALPHABET)
  })

  it('should update the state movesCount to the correct value', () => {
    const char = ALPHABET[Math.floor(Math.random() * ALPHABET.length)]
    // ALPHABET.splice(ALPHABET.indexOf(char), 1)
    wrapper.instance().onLetterClick(char)
    expect(wrapper).to.have.state('movesCount').to.deep.equal([1, 0])
  })

  it('should the state currentPlayer didn\'t change in singlePlayer Mode', () => {
    if (navigationMode == "singlePlayer") {
      const char = ALPHABET[Math.floor(Math.random() * ALPHABET.length)]
      // ALPHABET.splice(ALPHABET.indexOf(char), 1)
      wrapper.instance().onLetterClick(char)
      expect(wrapper).to.have.state('currentPlayer').to.deep.equal(0)
    }
  })

})
