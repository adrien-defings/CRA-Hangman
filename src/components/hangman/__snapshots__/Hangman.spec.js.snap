// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hangman /> should match with prev snapshot 1`] = `
<div
  id="gameContainer"
>
  <canvas
    height={95}
    id="canvas"
    width={95}
  >
    Sorry, this browser does not support &ltcanvas&gt.
  </canvas>
  <InfosJoueur
    color="primary"
    currentPlayer={0}
    isGameEnded={false}
    navigationMode="multiplayer"
    winner={3}
  />
  <span>
    0
     moves
  </span>
  <div
    id="word"
  >
    <span
      className="letter badge badge-light"
      key="0"
      letter="_"
    >
      _
    </span>
    <span
      className="letter badge badge-light"
      key="1"
      letter="_"
    >
      _
    </span>
    <span
      className="letter badge badge-light"
      key="2"
      letter="_"
    >
      _
    </span>
    <span
      className="letter badge badge-light"
      key="3"
      letter="_"
    >
      _
    </span>
  </div>
  <Letters
    color="primary"
    onClick={[Function]}
    usableLetters={
      Array [
        "A",
        "B",
        "C",
        "D",
        "E",
        "F",
        "G",
        "H",
        "I",
        "J",
        "K",
        "L",
        "M",
        "N",
        "O",
        "P",
        "Q",
        "R",
        "S",
        "T",
        "U",
        "V",
        "W",
        "X",
        "Y",
        "Z",
      ]
    }
  />
</div>
`;
